### PR TITLE
fix: add FeynmanFix upgrade for a testnet issue

### PR DESCRIPTION
**Pull request by Pythonberg1997**
### Description

This pr is to add FeynmanFix upgrade.

### Rationale

There's one issue within the `StakeHub` system contract. This pr is to add a new hardfork to upgrade the contract code.

For more info about the contract issue, please refer to https://github.com/bnb-chain/bsc-genesis-contract/pull/523 

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* add FeynmanFix upgrade

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -842,6 +853,20 @@ func (c *ChainConfig) IsOnFeynman(currentBlockNumber *big.Int, lastBlockTime uin
...
+	return c.IsLondon(num) && isTimestampForked(c.FeynmanFixTime, time)
+}
+
+// IsOnFeynmanFix returns whether currentBlockTime is either equal to the FeynmanFix fork time or greater firstly.
+func (c *ChainConfig) IsOnFeynmanFix(currentBlockNumber *big.Int, lastBlockTime uint64, currentBlockTime uint64) bool {
+	lastBlockNumber := new(big.Int)
+	if currentBlockNumber.Cmp(big.NewInt(1)) >= 0 {
+		lastBlockNumber.Sub(currentBlockNumber, big.NewInt(1))
+	}
+	return !c.IsFeynman(lastBlockNumber, lastBlockTime) && c.IsFeynman(currentBlockNumber, currentBlockTime)
```
Comment (forcodedancing) [inline params/config.go:867]: `IsFeynmanFix` here

File: core/systemcontracts/upgrade.go [hunk 2]
```diff
@@ -634,8 +636,8 @@ func init() {
...
+				CommitUrl:    "https://github.com/bnb-chain/bsc-genesis-contract/commit/a803125f635f315e332fd5164e91a3df1d48559c",
```
Comment (zjubfd) [inline core/systemcontracts/upgrade.go:639]: Please change the commit id of other contract code as well.

File: core/systemcontracts/upgrade.go [hunk 3]
```diff
@@ -664,6 +666,23 @@ func init() {
 			},
 		},
 	}
+
+	// This upgrade is to fix an error in testnet only. So the upgrade config of mainnet is empty.
+	feynmanFixUpgrade[chapelNet] = &Upgrade{
+		UpgradeName: "feynman",
```
Comment (zjubfd) [inline core/systemcontracts/upgrade.go:672]: The name should be `feynmanFix`

File: core/systemcontracts/upgrade.go [hunk 4]
```diff
@@ -664,6 +666,23 @@ func init() {
...
+			},
+		},
+	}
+
+	feynmanFixUpgrade[rialtoNet] = &Upgrade{
+		UpgradeName: "feynman",
```
Comment (zjubfd) [inline core/systemcontracts/upgrade.go:683]: ==> `feynmanFix`?

File: core/systemcontracts/upgrade.go [hunk 5]
```diff
@@ -664,6 +666,23 @@ func init() {
...
+			},
+		},
+	}
+
+	feynmanFixUpgrade[rialtoNet] = &Upgrade{
```
Comment (zjubfd) [inline core/systemcontracts/upgrade.go:682]: we should have this too?
```
	feynmanFixUpgrade[Mainnet] = &Upgrade{
		UpgradeName: "feynmanFix",
		Configs:     []*UpgradeConfig{},
	}
```

File: params/config.go (hunk 1 above)
Comment (Pythonberg1997) [reply params/config.go:867]: Fixed

File: core/systemcontracts/upgrade.go (hunk 2 above)
Comment (Pythonberg1997) [reply core/systemcontracts/upgrade.go:639]: Done

File: core/systemcontracts/upgrade.go (hunk 3 above)
Comment (Pythonberg1997) [reply core/systemcontracts/upgrade.go:672]: Fixed

File: core/systemcontracts/upgrade.go (hunk 4 above)
Comment (Pythonberg1997) [reply core/systemcontracts/upgrade.go:683]: Fixed

File: core/systemcontracts/upgrade.go (hunk 5 above)
Comment (Pythonberg1997) [reply core/systemcontracts/upgrade.go:682]: Done

**Comment by buddh0** [comment]
not support flag? maybe align with flag OverrideFeynman is better

File: core/systemcontracts/upgrade.go [hunk 6]
```diff
@@ -604,66 +606,88 @@ func init() {
 		Configs: []*UpgradeConfig{
 			{
 				ContractAddr: common.HexToAddress(ValidatorContract),
-				CommitUrl:    "https://github.com/bnb-chain/bsc-genesis-contract/commit/9d09d0a3d6332460a810188261ec8195e05aa218",
+				CommitUrl:    "https://github.com/bnb-chain/bsc-genesis-contract/commit/a803125f635f315e332fd5164e91a3df1d48559c",
```
Comment (zjubfd) [inline core/systemcontracts/upgrade.go:609]: The commit id need update

File: params/config.go [hunk 7]
```diff
@@ -459,12 +461,13 @@ type ChainConfig struct {
...
-	ShanghaiTime *uint64 `json:"shanghaiTime,omitempty"` // Shanghai switch time (nil = no fork, 0 = already on shanghai)
-	KeplerTime   *uint64 `json:"keplerTime,omitempty"`   // Kepler switch time (nil = no fork, 0 = already activated)
-	FeynmanTime  *uint64 `json:"feynmanTime,omitempty"`  // Feynman switch time (nil = no fork, 0 = already activated)
-	CancunTime   *uint64 `json:"cancunTime,omitempty"`   // Cancun switch time (nil = no fork, 0 = already on cancun)
-	PragueTime   *uint64 `json:"pragueTime,omitempty"`   // Prague switch time (nil = no fork, 0 = already on prague)
-	VerkleTime   *uint64 `json:"verkleTime,omitempty"`   // Verkle switch time (nil = no fork, 0 = already on verkle)
+	ShanghaiTime   *uint64 `json:"shanghaiTime,omitempty"`   // Shanghai switch time (nil = no fork, 0 = already on shanghai)
+	KeplerTime     *uint64 `json:"keplerTime,omitempty"`     // Kepler switch time (nil = no fork, 0 = already activated)
+	FeynmanTime    *uint64 `json:"feynmanTime,omitempty"`    // Feynman switch time (nil = no fork, 0 = already activated)
+	FeynmanFixTime *uint64 `json:"feynmanFixTime,omitempty"` // Feynman switch time (nil = no fork, 0 = already activated)
```
Comment (forcodedancing) [inline params/config.go:467]: Fix comment `Feynman switch time`

**Review by zjubfd — APPROVED** [review]

**Review by forcodedancing — APPROVED** [review]

**Pull request merged**